### PR TITLE
refactoring: add FieldString

### PR DESCRIPTION
**Pull request by soenkehahn**

### CODE REVIEW

File: src/System/Console/GetOpt/Generics/FieldString.hs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+normalized :: FieldString -> String
+normalized (FieldString s) = normalize s
+
+mkFieldString :: String -> FieldString
+mkFieldString = FieldString
+
+matches :: String -> FieldString -> Bool
+matches s field =
+  s == normalized field ||
+  normalize s == normalized field
```
Comment (tuleism) [inline src/System/Console/GetOpt/Generics/FieldString.hs:22]: We only need the 2nd condition here right?

File: src/System/Console/GetOpt/Generics/Modifier.hs [hunk 2]
```diff
@@ -64,51 +64,49 @@ data Modifiers = Modifiers {
...
+mkModifiers :: [Modifier] -> Modifiers
+mkModifiers = foldl' inner empty
   where
     empty :: Modifiers
     empty = Modifiers [] [] [] [] Nothing
 
-    inner :: Modifiers -> Modifier -> Result Modifiers
+    inner :: Modifiers -> Modifier -> Modifiers
     inner (Modifiers shorts renamings args help version) modifier = case modifier of
       (AddShortOption option short) ->  do
```
Comment (tuleism) [inline src/System/Console/GetOpt/Generics/Modifier.hs]: We don't need `do` anymore.

File: src/System/Console/GetOpt/Generics/FieldString.hs (hunk 1 above)
Comment (soenkehahn) [inline src/System/Console/GetOpt/Generics/FieldString.hs:22]: Yes, of course, changing.

File: src/System/Console/GetOpt/Generics/Modifier.hs [hunk 3]
```diff
@@ -64,51 +64,49 @@ data Modifiers = Modifiers {
...
+        Modifiers shorts renamings args (insert option helpText help) version
       (AddVersionFlag v) -> do
-        return $ Modifiers shorts renamings args help (Just v)
+        Modifiers shorts renamings args help (Just v)
 
-mkShortOptions :: Modifiers -> String -> [Char]
-mkShortOptions (Modifiers shortMap _ _ _ _) option =
-    fromMaybe [] (lookup option shortMap)
+lookupMatching :: [(String, a)] -> FieldString -> Maybe a
+lookupMatching list option = fmap snd $ listToMaybe $ filter (\ (from, _) -> from `matches` option) list
```
Comment (tuleism) [inline src/System/Console/GetOpt/Generics/Modifier.hs]: `listToMaybe $ filter p` == `find p`

File: src/System/Console/GetOpt/Generics/Modifier.hs (hunk 2 above)
Comment (soenkehahn) [inline src/System/Console/GetOpt/Generics/Modifier.hs]: Good catch!

File: src/System/Console/GetOpt/Generics/Modifier.hs (hunk 3 above)
Comment (soenkehahn) [inline src/System/Console/GetOpt/Generics/Modifier.hs]: Changing.

File: test/System/Console/GetOpt/Generics/FieldStringSpec.hs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+      normalize "foo2bar" `shouldBe` "foo2bar"
+
+    it "when there's one valid character it returns only dashes and lower case characters" $ do
+      property $ \ x ->
+        any isValidInputChar x ==>
+        normalize x `shouldSatisfy`
+          (\ s -> all isAllowedOutputChar s)
+
+    it "when there are no valid characters it returns its input" $ do
+      property $ \ (filter (not . isValidInputChar) -> x) ->
```
Comment (tuleism) [inline test/System/Console/GetOpt/Generics/FieldStringSpec.hs]: I think it's more readable with

``` haskell
property $ \ x
  all (not . isValidInputChar) x ==>
```

**Comment by tuleism** [comment]
LGTM :+1:

**Pull request merged**